File: client/common/protocol/winners_response_message.go
package protocol

import (
	"encoding/binary"
	"errors"
)

type WinnersResponse struct {
	winners []string
}

func (wr *WinnersResponse) ShouldAck() bool {
	return false
}

func (wr WinnersResponse) addField(stream *[]byte, field string) {
	*stream = append(*stream, byte(len(field)))

	*stream = append(*stream, []byte(field)...)
}

func (wr *WinnersResponse) Serialize() []byte {
	body := make([]byte, 0)
	for _, document := range wr.winners{
		wr.addField(&body, document)
	}
	header := buildHeader([]byte{WINNRESP_OP}, len(body))
	return append(header, body...)
}

func (wr WinnersResponse) deserializeFieldLength(stream *[]byte) (int, error) {
	if len(*stream) == 0 {
		return 0, errors.New("stream is shorter than expected")
	}
	fieldSize := []byte{0, (*stream)[0]}
	length := binary.BigEndian.Uint16(fieldSize)
	
	*stream = (*stream)[1:]

	return int(length), nil
}

func (wr WinnersResponse) deserializeField(fieldLength int, stream *[]byte) (string, error) {
	field := ""
	if len(*stream) < fieldLength {
		return field, errors.New("stream is shorter than expected")
	}

	field = string((*stream)[:fieldLength])

	*stream = (*stream)[fieldLength:]

	return field, nil
}


func (wr WinnersResponse) getFieldFromStream(stream *[]byte) (string, error) {
	fieldLength, err := wr.deserializeFieldLength(stream)
	field := ""
	if err == nil {
		field, err = wr.deserializeField(fieldLength, stream)
	}

	return field, err

}

func (wr *WinnersResponse) Deserialize(stream []byte) error {
	err := checkHeader(stream, WINNRESP_OP)

	if err != nil {
		return err
	}

	wr.winners = nil
	stream = stream[HEADER_SIZE:]
	for len(stream) > 0 {
		field, err := wr.getFieldFromStream(&stream)
		if err != nil {
			return err
		}
		wr.winners = append(wr.winners, field)
	}

	return nil
}

File: server/common/protocol/winners_response_message.py
from common.protocol.message import Message
from common.utils import Bet


class WinnersResponseMessage(Message):
    WINNRESP_OP = bytes([0x07])

    def __init__(self, bets: list[Bet] = []) -> None:
        self.winners = list(map(lambda bet: bet.document, bets))
    
    def __eq__(self, __value: object) -> bool:
        try:
            result = all(map(lambda x: x[0]==x[1], zip(self.winners, __value.winners)))
        except Exception:
            result = False
        return result
            

    def __serialize_document(self, document: str) -> bytes:
        serialized_document = bytes([len(document)])
        serialized_document += document.encode()
        return serialized_document

    def serialize(self) -> bytes:
        body = bytes()

        for document in self.winners:
            body += self.__serialize_document(document)

        return self._build_header(self.WINNRESP_OP, len(body)) + body
    
    def __deserialize_field(self, field_length: int, stream: bytes) -> (str, int):
        if len(stream) < field_length:
            
            return "", -1
        
        field = stream[:field_length].decode()
        return field, field_length

    def __deserialize_field_length(self, stream: bytes) -> int:
        if len(stream) == 0:
            return -1
        return int.from_bytes(stream[:1], self.ENDIAN, signed=False)

    def __get_field_from_stream(self, stream: bytes) -> (str, int):
        field_length = self.__deserialize_field_length(stream)
        if field_length < 0:
            
            return "", field_length
        stream = stream[1:]
        field, err = self.__deserialize_field(field_length, stream)
        return field, err
    
    def deserialize(self, stream: bytes) -> bool:
        if not self._check_header(stream, self.WINNRESP_OP):
            return False

        stream = stream[self.HEADER_SIZE:]

        while len(stream) > 0:
            document, size = self.__get_field_from_stream(stream)
            if size < 0:
                return False
            self.winners.append(document)
            stream = stream[size:]

        return True

    def should_ack(self) -> bool:
        return False